리액트 8기 양승민 6주차 과제 제출

rukeni opened the pull request:
## 반영사항
- 반영 흐름
  - 강의를 따라서 생각에 흐름을 놓치지 않으려고 노력했습니다.
 
## 추후 반영예정
- 현재 react-router 버전에 맞춰 6버전으로 마이그레이션 할 예정입니다.

## 왜 그렇게 수행했는지
- 과제를 하는 것도 중요하지만 생각하는 방식을 잘 기억하려고 노력했습니다.

## 테스트
```npm run lint``` 통과
```npm run test``` reducer line 54번 줄 빼고 통과

rukeni committed:
Chore: @swc/jest 설치

rukeni committed:
Refactor: App에서 RestaurantsPage 컴포넌트로 분리하여 page단위로 만든다

rukeni committed:
Chore: webpack devServer historyApiFallback 설정

rukeni committed:
Add: Router별 페이지 분리 및 테스트 코드 추가

1. AboutPage, HomePage, NotFoundPage 추가
2. Page별 렌더만 하는 테스트 코드 추가
3. App에서 간단한 Router추가

rukeni committed:
Chore: react-router-dom 설치

rukeni committed:
Refactor: react-router 적용

rukeni committed:
Refacotor: 테스트를 위해 <BrowserRouter> index.js로 이동

rukeni committed:
Test: 각 라우터 별로 렌더링 되는지 테스트한다.

1. HomePage
2. AboutPage
3. RestaurantsPage
4. NotFoundPage

rukeni committed:
Test: Link태그가 동작되도록 MemoryRouter 테스트 코드에 추가

rukeni committed:
Modify: 테스트가 되도록 About페이지 문구 수정

rukeni committed:
Add: 링크타고 HomePage로 갈 수 있는 헤더 생성

hannut91 commented on webpack.config.js:
```diff
@@ -14,4 +14,9 @@ module.exports = {
   resolve: {
     extensions: ['.js', '.jsx'],
   },
+  devServer: {
+    historyApiFallback: {
+      index: 'index.html',
+    },
+  },
```
이 문서도 한 번 읽어보세요

https://github.com/bripkens/connect-history-api-fallback#introduction

hannut91 commented:
이번주도 화이팅입니다!

hannut91 commented:
기다리고 있습니다!

rukeni committed:
Refactor: 폴더구조 개편

각각 관심사에 맞게 파일 구조를 개편하였습니다.

1. 페이지만 담당한 pages폴더
2. 기능들을 관심사별로 분리하면 features폴더
3. 전역에 영향을 끼치는 기능을 모아놓은 apps폴더
4. 서비스로직이 존재하는 services폴더

rukeni committed:
Refactor: store폴더 만들어서 redux를 따로 분리

rukeni committed:
Test: UI부분을 Categories컴포넌트로 분리한다.

rukeni committed:
Add: 컨테이너에서 Categories로 분리

rukeni committed:
Test: 컨테이너에서 UI컴포넌트를 분리한다.

rukeni committed:
Add: Regions 컴포넌트로 UI 분리

rukeni committed:
Fix: import 잘못된 경로 수정

rukeni committed:
Test: Restaurants 컨테이너에서 UI를 분리한다.

rukeni committed:
Refactor: Restaurants로 UI 분리

rukeni committed:
Fix: 잘못된 경로 정정

rukeni commented:
## 현재 반영사항
- 리팩토링 흐름
  - 새롭게 폴더 구조를 잡아보았습니다.
  - UI컴포넌트와 데이터를 주입하는 컨테이너를 나누었습니다.


## 왜 그렇게 수행했는지
- 관심사별로 구조를 개편하고 레이어별로 구조를 나누고 싶었습니다.
- UI 컴포넌트는 잦은 수정이 있을 수 있고 데이터 흐름과는 무관한 레이어라고 판단했습니다.
- 그렇기 때문에 UI 컴포넌트는 자체적으로 순수하게 유지하고 싶었고 그게 추후, 비주얼 테스트를 포함한 단위테스트에도 유리하다고 판단했습니다.

## 테스트
```npm run lint``` 통과
```npm run test``` reducer line 54번 줄 빼고 통과

rukeni committed:
Chore: react-router 버전 6으로 업그레이드

rukeni committed:
Refactor: react-router 6버전에 맞게 문법 수정

rukeni commented:
## 현재 반영사항
- react-router를 5버전에서 6버전으로 업그레이드 하였습니다.

## 왜 그렇게 수행했는지
- 강의의 흐름을 놓치고 싶지 않아서 강의와 똑같은 5버전으로 설치하였습니다.
- 6버전으로 마이그레이션하는 연습을 염두했습니다.

## 테스트
```npm run lint``` 통과
```npm run test``` reducer line 54번 줄 빼고 통과

rukeni committed:
Refactor: 라우팅을 Router컴포넌트로 분리

rukeni committed:
Test: 헤더영역을 컴포넌트로 분리한다.

rukeni committed:
Add: 헤더 추가

rukeni commented:
## 현재 반영사항
- `App.jsx`의 라우팅 기능을 Router로 분리하였습니다.
- `App.jsx`의 헤더 영역을 `Header.jsx`로 분리하였습니다.

## 테스트
```npm run lint``` 통과
```npm run test``` reducer line 54번 줄 빼고 통과

rukeni commented:
아닛. 요구사항에 상세정보가 있었구나! 🧐  조속히 반영하겠습니다.

rukeni committed:
Add: 레스토랑 상세정보 fixture 추가

rukeni committed:
Add: api 요청 정보 추가

rukeni committed:
Fix: 경로 변경 및 잘못된 경로 수정

rukeni committed:
Add: 레스토랑 상세정보 mock 추가

rukeni committed:
Test: 레스토랑 ID가 있으면 레스토랑 상세정보를 가져온다.

rukeni committed:
Add: 레스토랑 상세 요청 추가

hannut91 commented on src/features/category/Categories.jsx:
```diff
@@ -0,0 +1,21 @@
+            onClick={() => onClick(id)}
+          >
+            {name}
+            {selectedCategory ? (
+              <>
+                {id === selectedCategory.id ? '(V)' : null}
+              </>
+            ) : null}
```
JSX에서 `false` , `true` , `undefined` , `null` 은 아무것도 렌더하지 않습니다. 따라서 3항 연산자를 사용해서 null을 반환하는 대신에 그냥 null을 반환해도 될 것 같습니다.

```jsx
            {selectedCategory && id === selectedCategory.id && (
              <>(V)</>
            )}
```

See also

- [https://reactjs.org/docs/jsx-in-depth.html#booleans-null-and-undefined-are-ignored](https://reactjs.org/docs/jsx-in-depth.html#booleans-null-and-undefined-are-ignored)

rukeni replied:
넵 감사합니다. 같은로직이 있는 `Regions.jsx`에도 반영하였습니다.

hannut91 reviewed (commented):
서버에서 정보를 올바르게 받지 못했을 경우도 처리를 해봅시다

rukeni committed:
Refactor: 3항 연산자에서 &&연산자로 변경

rukeni committed:
Test: 레스토랑 상세정보에 들어오면 이름, 주소, 메뉴가 보여야 한다.

rukeni committed:
Add: Router에 레스토랑 상세정보 페이지 추가

rukeni committed:
Test: 레스토랑id에 맞는 상세정보를 보여줘야한다.

rukeni committed:
Add: RestaurantDetail 페이지 추가

rukeni committed:
Move: Restaurants -> Restaurants/Restaurants로 이동

rukeni committed:
Test: 레스토랑 ID를 이용하여 상세정보를 보여준다.

rukeni committed:
Add: Detail페이지 추가

rukeni committed:
Fix: webpack dev server가 절대경로를 못잡는 문제 수정

rukeni committed:
Add: Link태그 추가

rukeni committed:
Test: Link태그 추가를 위해 테스트코드에 MemoryRouter 추가

rukeni committed:
Test: MemoryRouter 테스트코드에 추가

rukeni committed:
Test: RestaurantDetail을 reducer에 반영 해야한다.

rukeni committed:
Add: 레스토랑 상세정보 적용하는 reducer 추가

rukeni committed:
Test: loadRestaurantDetail 테스트 코드 추가

1. selectedRestaurantId가 주어지면 setRestaurantDetail이 실행된다.

2. selectedRestaurantId가없으면 어떤 액션도 실행되지 않는다.

rukeni committed:
Add: action 함수 추가

1. loadRestaurantDetail
선택된 레스토랑이 있으면 api를 호출하고 없으면 null을 리턴한다.

2. setRestaurantDetail
레스토랑 디테일을 리턴한다.

rukeni committed:
Test: Link 컴포넌트를 위해 MemoryRouter 테스트코드에 추가

rukeni committed:
Modify: restaurantId를 매개변수에서 받도록 수정

rukeni committed:
Fix: fetchRestaurantById 실수로 구조분해로 인수받아오던 것을 일반 인수로 변경

rukeni committed:
Modify: 안쓰는 selectedRestaurantId 제거

rukeni committed:
Test: RestaurantDetail은 ID값에 맞는 레스토랑 상세정보를 표시해야한다.

rukeni committed:
Add: RestaurantDetail컴포넌트 기능 추가

rukeni committed:
Test: Detail페이지는 selector를 통해 레스토랑 상세정보를 가져온다.

rukeni committed:
Refactor: 테스트코드를 통과하도록 컴포넌트 기능 추가

rukeni committed:
Test: default reducer 테스트 추가

rukeni committed:
Fix: action이 없는 경우 optional처리

rukeni committed:
Test: restaurantDetail 값이 주어지지 않으면 렌더링 되지 않는다.

rukeni committed:
Fix: map함수 사용을 위해 값이 아직 없을때 배열 일 수 있도록 빈 배열로 초기화

rukeni committed:
Fix: 문구수정

rukeni committed:
Rename: 컴포넌트이름 변경

1. Restaurant기능이라는 공통 관심사하에 상세정보 컴포넌트들을 restaurant폴더로 모았다.
2. DetailContainer를 만들면서 View역할을 하는 컴포넌트를 Detail컴포넌트로 바꾸었다.
3. Page 컴포넌트 이름을 동일해서 해깔리지 않을까 고민했는데 바꾸지 않고 검색에만 용이하게 export default에 Page라는 접미사만 두었다.

rukeni committed:
Test: isLoading을 변경한다.

rukeni committed:
Add: setIsLoading 리듀서 추가

rukeni committed:
Add: setIsLoading 액션함수 추가

rukeni committed:
Test: 상태 변경에 대한 리듀서 함수 추가

1. setLoading은 isLoading을 true 또는 false로 바꾸고 isError와 errorMessage를 항상 초기화한다.

2. setError는 isError와 errorMessage를 변경하고 isLoading을 항상 false로 초기화한다.

rukeni committed:
Fix: 테스트에 통과할 수있도록 버그 수정

rukeni committed:
Test: restaurantDetail 관심사별로 loading, error, data를 처리할 수있도록 한다.

rukeni committed:
Refactor: 객체구조 변경에 따른 리듀서, 액션 수정

1. 객체구조가 restaurantDetail에서 restaurantDetail.data로 변경됨에 따라 접근자 코드를 수정하였다.

2. restaurantDetail을 찾아가기 위한 key를 action함수에 포함시켰다.

rukeni committed:
Test: restaurantDetail 객체 구조 변경에 따른 테스트 코드 수정

rukeni committed:
Refactor: 로딩 처리, 에러 처리 적용

rukeni committed:
Test: 객체 구조 변경에 따라 테스트 코드 수정

rukeni committed:
Refactor: 리듀서와 액션함수 수정

1. setLoading, setError가 첫번째 인자로 key를받아서 해당 프로퍼티의 loading, error를 변경할 수있도록 수정하였습니다.

2. setRestaurantDetail에서 데이터가 들어왔으므로 loading과,error를 초기화하도록 로직을 변경하였습니다.

rukeni committed:
Chore: @types/redux-mock-store 설치

rukeni committed:
Test: isLoading때와 isError일때 테스트 코드 추가

rukeni committed:
Test: loadRestaurantDetail()함수의 catch문 테스트코드 추가

rukeni committed:
Add: error가 발생하면 setError호출하여 isError를 true로 바꾼다.

rukeni commented on src/apps/store/actions.js:
```diff
@@ -67,3 +105,23 @@ export function loadRestaurants() {
+      const restaurantDetail = await fetchRestaurantById(restaurantId);
+      dispatch(setRestaurantDetail(restaurantDetail));
+    } catch (error) {
+      dispatch(setError('restaurantDetail', {
```
이부분의 테스트 커버리지가 계속 만족이 안되는데 어떻게해야 커버리지를 올릴 수 있을까요?

계속 삽질했는데 해결을 못해서 넘 신경쓰이네요. 🤣

왜 안될까..

일단 못다한 강아지 산책시키러 갔다가 내일 고민해보겠습니다.

rukeni commented:
> 서버에서 정보를 올바르게 받지 못했을 경우도 처리를 해봅시다

말씀해주신 에러에대해서 처리하였습니다. 저는 아래와 같은 방법으로 반영하였습니다.

## 현재 반영사항
- error 핸들링 로직을 추가하였습니다.
- loading 핸들링 로직을 추가하였습니다.

- 데이터를 fetch시작하면 `isLoading: true` 로 바뀝니다.
- fetch가 성공하면 `data`프로퍼티에 주입되면서 `isLoading: false`로 바뀝니다.
- fetch가 실패하면 `isError: true, errorMessage, isLoading: false` 로 처리됩니다.

- 각 status는 관심사마다 가지고 있고 구조예시는 다음과 같습니다.

```js

  restaurantDetail: {
    isLoading: false,
    isError: false,
    errorMessage: '',
    data: {},
  },


```

- 현재는 레스토랑 상세정보만 적용된 상태이며 나머지 요청은 내일 중으로 적용할 예정입니다!
- catch문에 있어서 그런건지 모르겠는데 커버리지 100을 못만들고 있습니다. 😥

## 테스트
```npm run lint``` 통과
```npm run test``` action.js line 121번 줄 빼고 통과

hannut91 commented:
지금 `../services/api` 모킹한 것을 보면

```js
export async function fetchRestaurantById() {
  return {};
}
```

위와 같이 빈 오브젝트를 반환하도록 모킹하고 있습니다. `fetch`를 모킹하더라도 애초에 `fetch`를 호출하지 않기 때문에 테스트가 실패하고 있네요.

그래서 위 코드를

```js
export const fetchRestaurantById = jest.fn({});
```

이렇게 바꾸고, 테스트 코드를

```js
beforeEach(() => {
  fetchRestaurantById.mockRejectedValue(new Error('error'));
});
```

이렇게 바꾸어야 합니다. 그리고 기타 등등 틀린게 있는데, 그건 수정하실 수 있을 것 같아요.

rukeni commented:
> 지금 `../services/api` 모킹한 것을 보면
> 
> ```js
> export async function fetchRestaurantById() {
>   return {};
> }
> ```
> 
> 위와 같이 빈 오브젝트를 반환하도록 모킹하고 있습니다. `fetch`를 모킹하더라도 애초에 `fetch`를 호출하지 않기 때문에 테스트가 실패하고 있네요.
> 
> 그래서 위 코드를
> 
> ```js
> export const fetchRestaurantById = jest.fn({});
> ```
> 
> 이렇게 바꾸고, 테스트 코드를
> 
> ```js
> beforeEach(() => {
>   fetchRestaurantById.mockRejectedValue(new Error('error'));
> });
> ```
> 
> 이렇게 바꾸어야 합니다. 그리고 기타 등등 틀린게 있는데, 그건 수정하실 수 있을 것 같아요.

헛... ㅇㅅㅇ. 감사합니다! 그것때문에 어제 잠이 안왔는데 포인트는 가까운데 있군요.

기타 등등 틀린거는 무엇일까요. 🤔 하루동안 한번 잡아보겠습니다.

rukeni committed:
Test: fetchRestaurantById mock함수 jest.fn()으로 변경

rukeni committed:
Test: fetchRestaurantById mock 함수 변경에 따른 테스트 코드 추가

1. loadRestaurantDetail을 테스트할때 resolve로 restaurantDetail이 들어오면 제대로 액션에 반환되는 지 확인

2. loadRestaurantDetail이 에러일때, Promise를 reject시키고 호출한 액션이 setError를 호출한 결과와 같은지 확인한다.

rukeni committed:
Fix: 에러 메시지를 받도록 수정

rukeni committed:
setError함수 수정

1. setError는 에러를 셋팅하는 함수라서 isError가 항상 true이다. 따라서 따로 인수로 받을 필요가 없다.

2. isError가 없어지면서 객체형식으로 받을 필요가 없어졌다.

rukeni committed:
Modify: restaurants 구조로 수정

1. 객체로 변경하여 loading과 error를 처리한다.
2. 기존정보는 data프로퍼티에 배열로 담는다.

rukeni committed:
Test: restaurants 구조 변경에 따른 테스트 코드 수정

rukeni committed:
Modify: restaurants 구조 변경에 따라 액션, 리듀서 수정

- isLoading, isError, errorMessage, key값 추가 반영

rukeni committed:
Test: fetchRestaurants mock 함수 jest.fn()으로 변경

rukeni committed:
Test: restaurants 객체 구조 변경에 따른 테스트 코드 수정

rukeni committed:
Test: fetchRestaurants reject시 error 상태를 갱신한다.

rukeni committed:
Refactor: isLoading, isError, errorMessage RestaurantsContainer컴포넌트에 반영

rukeni committed:
Test: selector로 가져온 restaurants 객체 데이터 변경에 따른 테스트 코드 수정

rukeni committed:
Test: loading일때, error일때 테스트코드 보강

rukeni committed:
Modify: regions, categories state에 loading, error 구조 반영

rukeni committed:
Test: catch 테스트를 위해 fetchRegions와 fetchCategories를 jest.fn()으로 변경한다.

rukeni committed:
Test: regions, categories 객체 구조 변경에 따른 테스트코드 수정

rukeni committed:
Modify: loadInitialData에 로딩과 에러처리 추가

rukeni committed:
Test: loadInitialData 테스트 코드 추가 및 변경

1. resolve일때 mockStore와 mockResolvedValue 테스트 코드에 추가

2. reject일때 테스트 코드 추가

rukeni committed:
Refactor: RegionsContainer에 loading, error 반영

rukeni committed:
Test: 로딩중일때와 에러있을때에 대한 테스트 코드 추가

rukeni committed:
Test: 객체구조 변경에 따른 mockStore 수정

rukeni committed:
Refactor: CategoriesContainer에 loading과 error처리 반영

rukeni committed:
Test: loading과 error 상황에 대한 테스트 코드 추가

rukeni committed:
Modify: 데이터 구조 변경에 따른 selectedRegion, selectedCategory 타겟 변경

rukeni committed:
Test: mockState 구조 수정

rukeni commented:
제가 원하던대로 thunk가 에러와 로딩을 처리하게되었습니다. 근데 리듀서와 액션 사이즈가 커지면서 코드가 장황해질 기미를 보이기 시작했습니다.  🤔

hannut91 commented on src/App.test.jsx:
```diff
@@ -1,34 +1,78 @@
+        errorMessage: '',
+        data: [{ name: '서울', id: 1 }],
+      },
+      categories: {
+        isLoading: false,
+        isError: false,
+        errorMessage: '',
+        data: [],
+      },
+      restaurants: {
+        isLoading: false,
+        isError: false,
+        errorMessage: '',
+        data: [],
+      },
```
각 데이터마다 상태를 만들어주셨군요. 그래서 상태에 따라 다른 것을 보여줄 수 있겠네요.

그런데 점점 이런게 많아질수록 고통스러워 지겠는데요. 묘수가 있을까요?

rukeni replied:
흠 주말에 부모님댁에 다녀와서 코딩을 못했습니다 ^^;

일단 action 함수와 reducer도 관심사 별로 분리는 필요한 것 같습니다.

`restaurants/restaurants.reducer.js` 라는 형식으로 도메인에 따라 분리하는게 나을지, `reducers/restaurants.reducer.js`라는 형식으로 담당하는 레이어에 따라 분리하는게 나을지는 고민이 듭니다.

다음 주 주제에서는 이 부분을 고려하여 작업해보겠습니다. 

항상 일깨움을 주는 물음을 주셔서 감사합니다.

hannut91 commented on src/apps/store/actions.js:
```diff
@@ -0,0 +1,156 @@
+      dispatch(setLoading('regions', true));
+      dispatch(setLoading('categories', true));
+      // TODO: loadInitialData처리 마무리
+      const regions = await fetchRegions();
+      dispatch(setRegions(regions));
+
+      const categories = await fetchCategories();
+      dispatch(setCategories(categories));
```
지금은 지역목록을 가져오고 나서 카테고리 목록을 가져오고 있는 것 같아요. 이 데이터는 서로 연관관계가 없으므로, 동시에 가져오도록 하는 것은 어떨까요?
